app/models/concerns/labware_creators/donor_pooling_validator.rb

```diff
@@ -0,0 +1,121 @@
+  #
+  # @return [void]
+  def source_plates_must_exist
+    return if source_plates.size == minimal_barcodes.size
+
+    errors.add(:source_plates, SOURCE_PLATES_MUST_EXIST)
+  end
+
+  # Validates that the number of calculated pools does not exceed the configured
+  # number of pools. If the number of calculated pools is greater, an error is
+  # added to the :source_plates attribute.
+  #
+  # @return [void]
+  def number_of_pools_must_not_exceed_configured
+    # Don't add this error if there are already errors about missing donor_ids.
+    invalid_wells_hash = locations_with_missing_donor_id(source_plates)
+    return if invalid_wells_hash.any?
+
+    return if pools.size <= number_of_pools
+
+    errors.add(:source_plates, format(NUMBER_OF_POOLS_MUST_NOT_EXCEED_CONFIGURED, pools.size, number_of_pools))
+  end
+
+  # Validates that all wells with aliquots must have a donor_id.
+  # It uses the locations_with_missing_donor_id method to find any wells that are
+  # missing a donor_id. If any such wells are found, it adds an error message to
+  # the source_plates attribute, formatted with the barcodes of the plates and
+  # the wells that are missing a donor_id.
+  #
+  # @return [void]
+  def wells_with_aliquots_must_have_donor_id
+    invalid_wells_hash = locations_with_missing_donor_id(source_plates)
+    return if invalid_wells_hash.empty?
+
+    formatted_string = invalid_wells_hash.map { |barcode, locations| "#{barcode}: #{locations.join(', ')}" }.join(' ')
+    errors.add(:source_plates, format(WELLS_WITH_ALIQUOTS_MUST_HAVE_DONOR_ID, formatted_string))
+  end
+
+  private
+
+  # Checks each well in each source plate for missing donor_id. Returns a hash
+  # with keys as the barcodes of source plates and values as arrays of well
+  # locations with missing donor_id. If a plate has no wells with missing
+  # donor_id, it is not included in the returned hash. This method is used by
+  # the wells_with_aliquots_must_have_donor_id method to generate an error message.
+  #
+  # @param source_plates [Array] An array of source plates to check.
+  # @return [Hash] A hash mapping source plate barcodes to arrays of invalid wells.
+  def locations_with_missing_donor_id(source_plates)
+    source_plates.each_with_object({}) do |source_plate, hash|
+      invalid_wells = source_plate.wells.select { |well| missing_donor_id?(well) }
+      hash[source_plate.human_barcode] = invalid_wells.map(&:location) if invalid_wells.any?
+    end
+  end
+
+  # Checks if a well is missing a donor_id. Returns false if the well has not
+  # passed or if there is no aliquot. If there is an aliquot, it checks if the
+  # associated sample_metadata has a donor_id. If the donor_id is missing, it
+  # returns true. Otherwise, it returns false.
+  #
+  # @param well [Well] The well to check.
+  # @return [Boolean] True if the well is missing a donor_id, false otherwise.
+  def missing_donor_id?(well)
+    return false unless well.passed?
```

Comment:
You've got this check on well state here and also in the labware creator (`labware_wells` method, and also `WellFilter`)? Would be cleaner for this method to only receive pre-checked wells. In case we decide on different criteria for state in the future, would be safer if the check was in only one place.

Reply:
Thanks. That is right. I have updated the code to use the same pre-checked wells.